concordium-std/src/traits.rs

```diff
@@ -114,34 +118,78 @@ pub trait HasReceiveContext<Error: Default = ()>: HasCommonData {
     fn owner(&self) -> AccountAddress;
 }
 
-/// A type that can serve as the contract state type.
-pub trait HasContractState<Error: Default = ()>
+/// A type that can serve as the contract state entry type.
+pub trait HasContractStateEntry
 where
     Self: Read,
-    Self: Write<Err = Error>,
-    Self: Seek<Err = Error>, {
-    type ContractStateData;
-    /// Open the contract state. Only one instance can be opened at the same
-    /// time.
-    fn open(_: Self::ContractStateData) -> Self;
+    Self: Write<Err = Self::Error>,
+    Self: Seek<Err = Self::Error>, {
+    type StateEntryData;
+    type StateEntryKey;
+    type Error: Default;
+
+    fn open(_: Self::StateEntryData, _: Self::StateEntryKey, entry_id: StateEntryId) -> Self;
 
-    /// Get the current size of contract state.
+    /// Get the current size of the entry.
     fn size(&self) -> u32;
 
-    /// Truncate the state to the given size. If the given size is more than the
-    /// current state size this operation does nothing. The new position is at
+    /// Truncate the entry to the given size. If the given size is more than the
+    /// current size this operation does nothing. The new position is at
     /// most at the end of the stream.
     fn truncate(&mut self, new_size: u32);
 
     /// Make sure that the memory size is at least that many bytes in size.
     /// Returns true iff this was successful. The new bytes are initialized as
     /// 0.
     fn reserve(&mut self, len: u32) -> bool;
+
+    fn get_key(&self) -> Vec<u8>;
+}
+
+pub trait HasContractStateLL {
+    type ContractStateData;
+    type EntryType: HasContractStateEntry;
+    type IterType: Iterator<Item = Self::EntryType>;
+
+    /// Open the contract state. Only one instance can be opened at the same
+    /// time.
+    fn open(_: Self::ContractStateData) -> Self;
+
+    /// Get an entry in the state.
+    fn entry(&mut self, key: &[u8]) -> EntryRaw<Self::EntryType>;
+
+    /// Lookup an entry in the state.
+    fn lookup(&self, key: &[u8]) -> Option<Self::EntryType>;
+
+    /// Delete an entry.
+    /// Returns whether the entry actually existed.
+    /// The only way this will return false is if the entry was deleted via
+    /// `delete_prefix`.
+    fn delete_entry(&mut self, entry: Self::EntryType) -> bool;
+
+    /// If exact is set, delete the specific key. Otherwise, delete the entire
+    /// subtree. Returns whether the entry or subtree existed.
+    fn delete_prefix(&mut self, prefix: &[u8], exact: bool) -> bool;
+
+    /// Get an iterator over a map in the state.
+    fn iterator(&self, prefix: &[u8]) -> Self::IterType;
+}
+
+pub trait Persistable<S>
+where
+    Self: Sized, {
+    // TODO: Should this return something?
+    fn store(self, prefix: &[u8], state_ll: Rc<RefCell<S>>);
```

Comment:
As we discussed, it would be good to get rid of this `Rc<RefCell<S>>` for the host backed implementation.